wire-ios/Wire-iOS/Sources/UserInterface/Calling/Analytics/CallEndedAnalyticsController.swift

```diff
@@ -0,0 +1,288 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireAnalytics
+import WireLogging
+import WireSyncEngine
+
+final class CallEndedAnalyticsController<CallCenter: WireCallCenterV3> {
+
+    private let contextProvider: ContextProvider
+    private let notificationCenter: NotificationCenter
+    private let analyticsEventTracker: () -> (any AnalyticsEventTracker)?
+
+    private var eventInfo: EventInfo?
+    private var callStateObservationToken: AnyObject?
+    private var toggleVideoObservationToken: AnyObject?
+    private var callParticipantObsererToken: AnyObject?
+
+    private let logger: LoggerProtocol
+    private let currentDateProvider: CurrentDateProviding
+
+    init(
+        contextProvider: ContextProvider,
+        notificationCenter: NotificationCenter,
+        analyticsEventTracker: @escaping () -> (any AnalyticsEventTracker)?,
+        logger: LoggerProtocol,
+        currentDateProvider: CurrentDateProviding
+    ) {
+        self.contextProvider = contextProvider
+        self.notificationCenter = notificationCenter
+        self.analyticsEventTracker = analyticsEventTracker
+        self.logger = logger
+        self.currentDateProvider = currentDateProvider
+
+        setupObservations()
+    }
+
+    deinit {
+        let notificationCenter = notificationCenter
+        if let toggleVideoObservationToken {
+            notificationCenter.removeObserver(toggleVideoObservationToken)
+        }
+    }
+
+    private func setupObservations() {
+
+        callStateObservationToken = CallCenter.addCallStateObserver(
+            observer: self,
+            contextProvider: contextProvider
+        )
+
+        toggleVideoObservationToken = notificationCenter.addObserver(
+            forName: WireCallCenterV3.didToggleVideoNotification, object: .none, queue: .none
+        ) { [weak self] notification in
+            self?.handleToggleVideoNotification(notification)
+        }
+    }
+
+    private func handleIncomingCall(
+        _ conversation: ZMConversation,
+        _ isVideoCall: Bool
+    ) {
+        if let eventInfo {
+            logger.error("handleIncomingCall: expected eventInfo to be nil, but is: \(eventInfo)")
+        }
```

Comment:
It's possible that you have an ongoing call (and therefore `eventInfo` should not be `nil`) and then receive another incoming call. Perhaps we can handle multiple call and event info pairs?